R/R/prophet.R

```diff
@@ -1274,9 +1288,134 @@ fit.prophet <- function(m, df, ...) {
 #' forecast <- predict(m, future)
 #' plot(m, forecast)
 #' }
-#'
 #' @export
 predict.prophet <- function(object, df = NULL, ...) {
+  
+  if (is.null(df)) {
+    m <- object
+    history.and.future <- m$history
+  } else {
+    if (nrow(df) == 0) {
+      stop("Dataframe has no rows.")
+    }
+    out <- setup_dataframe(object, df)
+    history.and.future <- out$df
+    m <- out$m
+  }
+  
+  out2 <- make_all_seasonality_features(m, history.and.future)
+  seasonal.features <- out2$seasonal.features
+  component.cols <- object$train.component.cols
+  
+  # Construct input to stan
+  dat <- list(
+    T = nrow(object$history),
+    K = ncol(seasonal.features),
+    S = length(object$changepoints.t),
+    y = object$history$y_scaled,                         
+    t = object$history$t,                                
+    cap = rep(0, nrow(object$history)),                         
+    t_change = array(object$changepoints.t),
+    X = as.matrix(seasonal.features[1:nrow(object$history),]),            
+    sigmas = array(out2$prior.scales),
+    tau = object$changepoint.prior.scale,
+    trend_indicator = as.numeric(object$growth == 'logistic'),
+    s_a = array(component.cols$additive_terms),
+    s_m = array(component.cols$multiplicative_terms),
+    
+    T_pred = nrow(history.and.future),
+    t_pred = history.and.future$t,
+    cap_pred = 
+      if(class(history.and.future$cap) == "NULL") { 
+        rep(0, length(history.and.future$t)) 
+      } else { 
+        history.and.future$cap 
+      },
+    X_pred = as.matrix(seasonal.features),
+    n_samp = object$uncertainty.samples
+  )
+  
+  # Future # of change points assuming a homog. Poisson process
+  dat$S_pred <- stats::qpois(0.95, dat$S * (dat$T_pred - dat$T) / dat$T )
+  
+  # Run stan
+  if (exists(".prophet.stan.model")) {
+    model <- .prophet.stan.model
+  } else {
+    model <- get_prophet_stan_model()
+  }
+  
+  stan_init <- 
+    list(
+      k = object$params$k,
+      m = object$params$m,
+      delta = as.numeric(object$params$delta),
+      beta = as.numeric(object$params$beta),
+      sigma_obs = object$params$sigma_obs
+    )
+  
+  if (m$mcmc.samples > 0) {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = m$mcmc.samples
+    )
+    args <- utils::modifyList(args, list(...))
+    stan.fit <- do.call(rstan::sampling, args)
+  } else {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = 0,
+      as_vector = FALSE,
+      verbose = TRUE
```

Comment:
I expect we'll want `algorithm='Newton'` here too. There are reports that the L-BFGS can hang with very flat likelihood surfaces.

Reply:
Do you want to set it to 'Newton' and change the current behavior, where the user can choose between the algorithms while calling prophet?

Reply:
I meant in predict. In fit we'll keep it like it is (use L-BFGS by default and fallback to Newton), but in predict when we take our 1 step it can be with Newton.

Reply:
ok makes sense.